src/transaction/transaction_parser.c

```diff
@@ -46,11 +46,97 @@ bool parse_tx_fee_from_syscall(re_ins_syscall_t *syscall, uint256_t *tx_fee) {
         return false;
     }
 
+    if (syscall->call_data.data[0] != INS_SYSCALL_TX_FEE_RESERVE_PUT &&
+        syscall->call_data.data[0] != INS_SYSCALL_TX_FEE_RESERVE_TAKE) {
+        // Only these two values above are known to relate to TX fee.
+        return false;
+    }
+
     readu256BE(syscall->call_data.data + 1, tx_fee);
 
     return true;
 }
 
+/*
+ * Tx fee is calculated like so:
+ * Also a new syscall was added `FEE_RESERVE_TAKE (0x01)`, apart from old `FEE_RESERVE_PUT (0x00)`
+ * to be able to claim back unused fees in the fee reserve. A general user transaction will look
+ * like:
+ *
+ * DOWN <substate_id>
+ * SYSCALL FEE_RESERVE_PUT <put_amt>
+ * ...
+ * SYSCALL FEE_RESERVE_TAKE <take_amt0>
+ * ...
+ * SYSCALL FEE_RESERVE_TAKE <take_amt1>
+ * END
+ * SIG <signature>
+ *
+ * The corresponding fee paid in the transaction is then:
+ * `fee_paid = put_amt - sum(take_amt*)`
+ *
+ * There should only be one and only one `put_amt` value. There may be 0 or more `take_amt` values.
+ *
+ */
+static bool update_tx_fee_and_total_xrd_cost_if_needed(re_ins_syscall_t *syscall,
```

Comment:
@siy this function is new and probably the most notable and important update of this entire PR. It changes how tx fee is calculated.